Spigot-Server-Patches/0594-Use-TileEntity-s-BlockState-cache-when-ticking.patch

```diff
@@ -0,0 +1,73 @@
++++ b/src/main/java/net/minecraft/server/TileEntityTypes.java
+@@ -48,6 +48,8 @@ public class TileEntityTypes<T extends TileEntity> {
+     private final Set<Block> J;
+     private final Type<?> K;
+ 
++    private final Block validBlock; // Paper - Avoid using SingletonSets
++
+     @Nullable
+     public static MinecraftKey a(TileEntityTypes<?> tileentitytypes) {
+         return IRegistry.BLOCK_ENTITY_TYPE.getKey(tileentitytypes);
+@@ -60,29 +62,36 @@ public class TileEntityTypes<T extends TileEntity> {
+ 
+         Type<?> type = SystemUtils.a(DataConverterTypes.BLOCK_ENTITY, s);
+ 
+-        return (TileEntityTypes) IRegistry.a(IRegistry.BLOCK_ENTITY_TYPE, s, (Object) tileentitytypes_a.a(type));
++        return (TileEntityTypes) IRegistry.a(IRegistry.BLOCK_ENTITY_TYPE, s, tileentitytypes_a.a(type)); // Paper - Decompile Fix
+     }
+ 
+     public TileEntityTypes(Supplier<? extends T> supplier, Set<Block> set, Type<?> type) {
+         this.I = supplier;
+         this.J = set;
+         this.K = type;
++        // Paper start - Avoid using SingletonSets
++        if (set.size() == 1) {
++            this.validBlock = set.iterator().next();
++        } else {
++            this.validBlock = null;
++        }
++        // Paper end
+     }
+ 
+     @Nullable
+     public T a() {
+-        return (TileEntity) this.I.get();
++        return (T) this.I.get(); // Paper - Decompile Fix
+     }
+ 
+     public boolean isValidBlock(Block block) {
+-        return this.J.contains(block);
++        return this.validBlock == null ? this.J.contains(block) : this.validBlock.equals(block); // Paper - Avoid using SingletonSets
+     }
+ 
+     @Nullable
+     public T a(IBlockAccess iblockaccess, BlockPosition blockposition) {
+         TileEntity tileentity = iblockaccess.getTileEntity(blockposition);
+ 
+-        return tileentity != null && tileentity.getTileType() == this ? tileentity : null;
++        return tileentity != null && tileentity.getTileType() == this ? (T) tileentity : null; // Paper - Decompile Fix
+     }
+ 
+     public static final class a<T extends TileEntity> {
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index d91124bc85ed1a5b39e0ac3966f0af8ed9ee62b9..74251ddca4227260a5b53b6543e9b808163852db 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -799,7 +799,7 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+                             return String.valueOf(TileEntityTypes.a(tileentity.getTileType()));
+                         });
+                         tileentity.tickTimer.startTiming(); // Spigot
+-                        if (tileentity.getTileType().isValidBlock(this.getType(blockposition).getBlock())) {
++                        if (tileentity.getTileType().isValidBlock(tileentity.getBlock().getBlock())) { // Paper
```

Comment:
This check is specifically for ensuring world state, this change is basically invalid and would break this code from working, I feel that if we're going to do this, we might as well just hide this behind a system property, but, err, is this really that much of a saving grace?

Reply:
I'm unsure as to how much of an improvement this makes for most servers as all of the testing was done on a server with many tile entities. I never noticed any problems over a few days of testing which might be due to the TileEntity's cached BlockState possibly being updated by the time it reaches this validation check.

Reply:
it's for checking for state desync, e.g. block in the world was changed but the actual TE was somehow left dangling, this check invalidates that behavior totally outside of the first call to that method

Reply:
Yeah just gonna close this then considering that concern and how the removal of the SingletonSet probably would make next to no difference.